src/hotspot/share/gc/shenandoah/shenandoahConcurrentGC.cpp

```diff
@@ -150,15 +150,22 @@ bool ShenandoahConcurrentGC::collect(GCCause::Cause cause) {
     return false;
   }
 
+  assert(heap->is_concurrent_weak_root_in_progress(), "Must be doing weak roots now");
```

Comment:
This does not ring true when final mark is cancelled, or am I missing something?

Reply:
There is a cancellation check just above this line.

Reply:
Ah. Back when we wrote this code originally, we used GC state flags to figure out whether we really need to go into particular phases. But I guess cancellation check is OK too.